### PR TITLE
py/objtype: Implemented __delattr__ and __setattr__.

### PR DESCRIPTION
**Pull request by dmazzella**
This patch implement methods __delattr__ and __setattr__ for customize attribute access.

### CODE REVIEW

**Comment by dmazzella** [comment]
#2375

**Comment by dmazzella** [comment]
#2755 is not related ti this pr.

**Comment by dpgeorge** [comment]
Sorry but it might take some time to get back to this.

**Comment by dmazzella** [comment]
I am trying to implement the cache for call  `__delattr__`, `__getattr__ ` and `__setattr__`.

Before submitting a PR i want to discuss the goodness of my implementation:

into [objtype.h](https://github.com/micropython/micropython/blob/master/py/objtype.h#L35) i have added:

```
// cache __delattr__, __getattr__ and __setattr__ for efficient lookup of them
mp_map_t attrs_methods;
```

into [objtype.c](https://github.com/micropython/micropython/blob/master/py/objtype.c#L54) i have added: 

`mp_map_init(&o->attrs_methods, 0);`

into [objtype.c](https://github.com/micropython/micropython/blob/master/py/objtype.c#L536) i have added: 

```
        // if the requested attr is __setattr__ assign MP_OBJECT_NULL to dest[0] to indicate success 
        if (attr == MP_QSTR___setattr__) { 
            dest[0] = MP_OBJ_NULL; 
            return; 
        } 
        mp_obj_t dest_getattr[3];
        mp_map_elem_t *elem_getattr = mp_map_lookup(&self->attrs_methods, MP_OBJ_NEW_QSTR(MP_QSTR___getattr__), MP_MAP_LOOKUP_ADD_IF_NOT_FOUND);
        if (elem_getattr->value != MP_OBJ_NULL) {
            mp_seq_copy(dest_getattr, elem_getattr->value, 2, mp_obj_t);
        } else {
            mp_load_method_maybe(self_in, MP_QSTR___getattr__, dest_getattr);
            if (dest_getattr[0] != MP_OBJ_NULL) {
                elem_getattr->value = m_new(mp_obj_t, 2);
                mp_seq_copy(elem_getattr->value, dest_getattr, 2, mp_obj_t);
            }
        }

        if (dest_getattr[0] != MP_OBJ_NULL) {
            // __getattr__ exists, call it and return its result
            // XXX if this fails to load the requested attr, should we catch the attribute error and return silently?
            dest_getattr[2] = MP_OBJ_NEW_QSTR(attr);
            dest[0] = mp_call_method_n_kw(1, 0, dest_getattr);
            return;
        }
```

into [objtype.c](https://github.com/micropython/micropython/blob/master/py/objtype.c#L628) i have added: 

```
        // try __delattr__ first 
        if (attr != MP_QSTR___delattr__) {
            mp_obj_t dest_delattr[3];
            mp_map_elem_t *elem_delattr = mp_map_lookup(&self->attrs_methods, MP_OBJ_NEW_QSTR(MP_QSTR___delattr__), MP_MAP_LOOKUP_ADD_IF_NOT_FOUND);
            if (elem_delattr->value != MP_OBJ_NULL) {
                mp_seq_copy(dest_delattr, elem_delattr->value, 2, mp_obj_t);
            } else {
                mp_load_method_maybe(self_in, MP_QSTR___delattr__, dest_delattr);
                if (dest_delattr[0] != MP_OBJ_NULL) {
                    elem_delattr->value = m_new(mp_obj_t, 2);
                    mp_seq_copy(elem_delattr->value, dest_delattr, 2, mp_obj_t);
                }
            }
            if (dest_delattr[0] != MP_OBJ_NULL) {  
                // __delattr__ exists, call it and return its result  
                dest_delattr[2] = MP_OBJ_NEW_QSTR(attr);
                mp_call_method_n_kw(1, 0, dest_delattr);
                return true;  
            }
        }
```

and into [objtype.c](https://github.com/micropython/micropython/blob/master/py/objtype.c#L632) i have added: 

```
        // try __setattr__ first
        if (attr != MP_QSTR___setattr__) {
            mp_obj_t dest_setattr[4]; 
            mp_map_elem_t *elem_setattr = mp_map_lookup(&self->attrs_methods, MP_OBJ_NEW_QSTR(MP_QSTR___setattr__), MP_MAP_LOOKUP_ADD_IF_NOT_FOUND);
            if (elem_setattr->value != MP_OBJ_NULL) {
                mp_seq_copy(dest_setattr, elem_setattr->value, 2, mp_obj_t);
            } else {
                mp_load_method_maybe(self_in, MP_QSTR___setattr__, dest_setattr); 
                if (dest_setattr[0] != MP_OBJ_NULL) {
                    elem_setattr->value = m_new(mp_obj_t, 2);
                    mp_seq_copy(elem_setattr->value, dest_setattr, 2, mp_obj_t);
                }
            }
            if (dest_setattr[0] != MP_OBJ_NULL) { 
                // __setattr__ exists, call it and return its result 
                dest_setattr[2] = MP_OBJ_NEW_QSTR(attr); 
                dest_setattr[3] = value; 
                mp_call_method_n_kw(2, 0, dest_setattr); 
                return true; 
            }
        }
```

can be this patch acceptable?

Please give me a feedback

**Comment by pfalcon** [comment]
@dmazzella : I'm not sure about your questions in the previous comment. But I wanted to thank you for going thru the series of refactorings based on the previous comments, leading to this patch. It seems to be clean and understandable from a quick look. I actually hoped to play with it during NY holidays and then possibly merge or add further comments. I didn't get to it, and won't be able for some time likely, but please be sure to leave this PR in its current shape. Any further changes would need to be done on "top" of this patch, but separate from it (and then would need to wait until processing of this patch finishes).

**Comment by dmazzella** [comment]
@pfalcon the question is about the implementation of a cache for store `__delattr__`, `__getattr__` and `__setattr__` without a new PR only for see if is an acceptable solution.

**Comment by dpgeorge** [comment]
> // cache __delattr__, __getattr__ and __setattr__ for efficient lookup of them mp_map_t attrs_methods;

Instead of a mp_map_t for the cache, it would be much simpler (and more efficient) to have a fixed-size array with know offsets for the cached methods.  Eg a 3-element array, with the elements being get/set/del.

**Comment by dmazzella** [comment]
@dpgeorge 

i have used:

```
// cache __delattr__, __getattr__ and __setattr__ for efficient lookup of them
typedef struct _attrs_t {
    mp_obj_t __delattr__;
    mp_obj_t __getattr__;
    mp_obj_t __setattr__;
} attrs_t;

// instance object
// creating an instance of a class makes one of these objects
typedef struct _mp_obj_instance_t {
    mp_obj_base_t base;
    mp_map_t members;
    attrs_t attrs;
    mp_obj_t subobj[];
} mp_obj_instance_t;
```

and the lookup is like:

```
    // try __getattr__
    if (attr != MP_QSTR___getattr__) {
        // if the requested attr is __setattr__ assign MP_OBJECT_NULL to dest[0] to indicate success 
        if (attr == MP_QSTR___setattr__) { 
            dest[0] = MP_OBJ_NULL; 
            return; 
        } 
        mp_obj_t dest_getattr[3];
        if(self->attrs.__getattr__ != MP_OBJ_NULL) {
            memcpy(dest_getattr, &self->attrs.__getattr__, sizeof(mp_obj_t));
        } else {
            mp_load_method_maybe(self_in, MP_QSTR___getattr__, dest_getattr);
            if (dest_getattr[0] != MP_OBJ_NULL) {
                memcpy(&self->attrs.__getattr__, dest_getattr, sizeof(mp_obj_t));
            }
        }

        if (dest_getattr[0] != MP_OBJ_NULL) {
            // __getattr__ exists, call it and return its result
            // XXX if this fails to load the requested attr, should we catch the attribute error and return silently?
            dest_getattr[1] = self_in;
            dest_getattr[2] = MP_OBJ_NEW_QSTR(attr);
            dest[0] = mp_call_method_n_kw(1, 0, dest_getattr);
            return;
        }
    }
```

is an acceptable solution?

**Comment by dmazzella** [comment]
@dpgeorge  What is the downside of using the lookup like this:

into [objtype.c](https://github.com/micropython/micropython/blob/master/py/objtype.c#L322)
```
    mp_load_method_maybe(MP_OBJ_FROM_PTR(o), MP_QSTR___delattr__, o->attrs.__delattr__);
    mp_load_method_maybe(MP_OBJ_FROM_PTR(o), MP_QSTR___getattr__, o->attrs.__getattr__);
    mp_load_method_maybe(MP_OBJ_FROM_PTR(o), MP_QSTR___setattr__, o->attrs.__setattr__);
```

and into [objtype.c](https://github.com/micropython/micropython/blob/master/py/objtype.c#L535)

```
    // try __getattr__
    if (attr != MP_QSTR___getattr__) {
        // if the requested attr is __setattr__ assign MP_OBJECT_NULL to dest[0] to indicate success 
        if (attr == MP_QSTR___setattr__) { 
            dest[0] = MP_OBJ_NULL; 
            return; 
        } 

        if (self->attrs.__getattr__[0] != MP_OBJ_NULL) {
            // __getattr__ exists, call it and return its result
            // XXX if this fails to load the requested attr, should we catch the attribute error and return silently?
            self->attrs.__getattr__[2] = MP_OBJ_NEW_QSTR(attr);
            dest[0] = mp_call_method_n_kw(1, 0, self->attrs.__getattr__);
            return;
        }
    }
```

with a structure like: 

 ```
// cache __delattr__, __getattr__ and __setattr__ for efficient lookup of them
typedef struct _attrs_t {
    mp_obj_t __delattr__[3];
    mp_obj_t __getattr__[3];
    mp_obj_t __setattr__[4];
} attrs_t;

// instance object
// creating an instance of a class makes one of these objects
typedef struct _mp_obj_instance_t {
    mp_obj_base_t base;
    mp_map_t members;
    attrs_t attrs;
    mp_obj_t subobj[];
} mp_obj_instance_t;
```

?

**Comment by dmazzella** [comment]
@dpgeorge can be this PR a candidate for #2845 ?

File: py/mpconfig.h [hunk 1]
```diff
@@ -635,6 +635,12 @@ typedef double mp_float_t;
 #define MICROPY_PY_DESCRIPTORS (0)
 #endif
 
+// Whether to support object (__delattr__ and __setattr__)
+// This costs some code size and makes all load attrs and store attrs slow
+#ifndef MICROPY_PY_METHODS_DELATTRS_SETATTRS
+#define MICROPY_PY_METHODS_DELATTRS_SETATTRS (0)
```
Comment (dpgeorge) [inline py/mpconfig.h]: I'd call it simply `MICROPY_PY_DELATTR_SETATTR`.

File: py/objtype.c [hunk 2]
```diff
@@ -533,6 +533,14 @@ STATIC void mp_obj_instance_load_attr(mp_obj_t self_in, qstr attr, mp_obj_t *des
 
     // try __getattr__
     if (attr != MP_QSTR___getattr__) {
+        #if MICROPY_PY_METHODS_DELATTRS_SETATTRS
+        // if the requested attr is __setattr__ assign MP_OBJECT_NULL to dest[0] to indicate success
+        if (attr == MP_QSTR___setattr__) {
+            dest[0] = MP_OBJ_NULL;
```
Comment (dpgeorge) [inline py/objtype.c]: Is this line needed?  Tests seem to still pass without it.

**Comment by dpgeorge** [comment]
> What is the downside of using the lookup like this: ... done in this way the checking is done only once

Yes that certainly makes it efficient!  But note that now all instances increase by (3+3+4) words which is 40 bytes on a 32-bit machine.  That is a lot of extra RAM for each instance of a class that is created.  To improve on this you could just store the method and self, and then extract them to a temporary array (that's on the stack) when you do the call.  But still it will be 6 words extra per instance.

File: py/objtype.c (hunk 2 above)
Comment (dmazzella) [inline py/objtype.c]: without this:
```
  File "/home/dma/micropython/tests/basics/getattr.py", line 8, in __getattr__
RuntimeError: maximum recursion depth exceeded
```

Comment (pfalcon) [inline py/objtype.c]: @dmazzella : @dpgeorge talked about "line", not "lines". These fine distinctions are very important in programming.

**Comment by dmazzella** [comment]
I have semplified the structure for only store methods:
```
typedef struct _attrs_t {
    mp_obj_t __delattr__;
    mp_obj_t __getattr__;
    mp_obj_t __setattr__;
} attrs_t;
```

the lookup now is:

```
    mp_obj_t tmp_attr_method[4];
    mp_load_method_maybe(MP_OBJ_FROM_PTR(o), MP_QSTR___delattr__, tmp_attr_method);
    if (tmp_attr_method[0] != MP_OBJ_NULL) {
        o->attrs.__delattr__ = tmp_attr_method[0];
    }
    mp_load_method_maybe(MP_OBJ_FROM_PTR(o), MP_QSTR___getattr__, tmp_attr_method);
    if (tmp_attr_method[0] != MP_OBJ_NULL) {
        o->attrs.__getattr__ = tmp_attr_method[0];
    }
    mp_load_method_maybe(MP_OBJ_FROM_PTR(o), MP_QSTR___setattr__, tmp_attr_method);
    if (tmp_attr_method[0] != MP_OBJ_NULL) {
        o->attrs.__setattr__ = tmp_attr_method[0];
    }
```

delattr:

```
        // try __delattr__
        if (attr != MP_QSTR___delattr__) {
            if (self->attrs.__delattr__ != MP_OBJ_NULL) {
                // __delattr__ exists, call it and return its result
                mp_obj_t attr_delattr_method[3];
                attr_delattr_method[0] = self->attrs.__delattr__;
                attr_delattr_method[1] = self;
                attr_delattr_method[2] = MP_OBJ_NEW_QSTR(attr);
                mp_call_method_n_kw(1, 0, attr_delattr_method);
                return true;
            }
        }
```

getattr:

```
    // try __getattr__
    if (attr != MP_QSTR___getattr__) {
        // if the requested attr is __setattr__ assign MP_OBJECT_NULL to dest[0] to indicate success 
        if (attr == MP_QSTR___setattr__) {
            dest[0] = MP_OBJ_NULL;
            return; 
        }
        if (self->attrs.__getattr__ != MP_OBJ_NULL) {
            // __getattr__ exists, call it and return its result
            // XXX if this fails to load the requested attr, should we catch the attribute error and return silently?
            mp_obj_t attr_getattr_method[3];
            attr_getattr_method[0] = self->attrs.__getattr__;
            attr_getattr_method[1] = self;
            attr_getattr_method[2] = MP_OBJ_NEW_QSTR(attr);
            dest[0] = mp_call_method_n_kw(1, 0, attr_getattr_method);
            return;
        }
    }
```

setattr:

```
        // try __setattr__
        if (attr != MP_QSTR___setattr__) {
            if (self->attrs.__setattr__ != MP_OBJ_NULL) {
                // __setattr__ exists, call it and return its result
                mp_obj_t attr_setattr_method[4];
                attr_setattr_method[0] = self->attrs.__setattr__;
                attr_setattr_method[1] = self;
                attr_setattr_method[2] = MP_OBJ_NEW_QSTR(attr);
                attr_setattr_method[3] = value;
                mp_call_method_n_kw(2, 0, attr_setattr_method);
                return true;
            }
        }
```

how do you see it now?

**Comment by pfalcon** [comment]
> I have semplified the structure for only store methods:

Let me chime in: that won't work still, sorry. And the core of the problem is obvious: if the methods are defined on *class* level, they also should (could!) be cached on the *class* level, not repeatedly waste memory in each and every instance. I'm happy to report that I considered such caching and a lot of other cool optimizations in the first year of uPy being open-source. I'm less happy to report, but at least it's sad but true, that nowadays we aren't closer to such optimizations than we were 2-3 years ago. But instead we have a lot of down-to-earth, but important things to do, e.g. https://github.com/micropython/micropython/issues/1331 (I apologize in advanced if these don't interest you).

**Comment by dmazzella** [comment]
@pfalcon I worked on the "customize attribute access" because I came in handy directly implement it and thought It could be useful to other, my initial implementation was without cache (this PR), and now I try  to discussing about how optimize it.

If such changes are unnecessary I can close the PR and try to work on #1331

**Comment by pfalcon** [comment]
This PR, as it stands now, is needed, we appreciate the effort you put into it, and it will be merged soon (in some definition of "soon"), I set it's prio as high. Please concentrate on making (small) requested changes to it (see @dpgeorge's comments).

But I would recommend against any further work on "optimizing" it, because at the current state of affairs, it will be premature optimization: http://wiki.c2.com/?PrematureOptimization . Thanks for your consideration.

**Comment by dpgeorge** [comment]
Ok, this is finally merged in 18e65691661ef8e83060d0e72d66b16cc918c8b4.  I had to make some small changes to the code to match CPython, and I added some more tests (in the same file).  I also renamed the test file to class_delattr_setattr.py.

CPython does actually delegate a delete of `__delattr__` to the custom `__delattr__`, and also delegates a setattr of `__setattr__` to the custom `__setattr__`.  So I changed the code to allow this, simply by removing the unnecessary checks that the attr was one of these.

Thanks @dmazzella for persisting with this patch!

**Pull request closed without merging**